gslib/utils/text_util.py

```diff
@@ -263,99 +263,70 @@ def PrintableStr(input_val):
-      raise TypeError(
-        "'{}' is an invalid keyword argument for this function".format(key))
-    kw_params[key] = value
-  sep, end, file = kw_params.values()
-  if six.PY2:
-    if hasattr(file, 'mode') and 'b' not in file.mode and end == os.linesep:
-      end = b'\n'
-    pref_enc = 'utf-8'
-    if isinstance(sep, unicode):
-      sep = sep.encode(pref_enc)
-    if isinstance(end, unicode):
-      end = end.encode(pref_enc)
-    byte_objects = []
-    for object in objects:
-      if isinstance(object, str):
-        byte_objects.append(object)
-      elif isinstance(object, unicode):
-        byte_objects.append(object.encode(pref_enc))
+  def _get_args(**kwargs):
+    """Validates keyword arguments that would be used in Print
+
+    Valid keyword arguments, mirroring print(), are 'sep',
+    'end', and 'file'. These must be of types string, string,
+    and file / file interface respectively.
+
+    Returns the above kwargs of the above types.
+    """
+    expected_keywords = collections.OrderedDict([
+      ('sep', ' '),
+      ('end', '\n'),
+      ('file', sys.stdout)])
+
+    for key, value in kwargs.items():
+      if key not in expected_keywords:
+        error_msg = (
+          '{} is not a valid keyword argument. '
+          'Please use one of: {}')
+        raise KeyError(
+          error_msg.format(
+            key,
+            ' '.join(expected_keywords.keys())))
       else:
-        byte_objects.append(str(object).encode(pref_enc))
-    data = sep.join(byte_objects)
-    data += end
-    ttywrite(file, data)
-  else:  # PY3
-    def Py3Print(encoding, sep=sep, end=end):
-      """ Encode and send data to ttywrite with specified encoding.
-
-      Args:
-        encoding: Encoding to be used for all encode operations while printing.
-        sep: Line separator to be used for printing (defined in outer scope).
-        end: End character to be used for printing (defined in outer scope).
-
-      Returns: None
-      """
-      if isinstance(sep, str):
-        sep = sep.encode(encoding)
-      if isinstance(end, str):
-        end = end.encode(encoding)
-      byte_objects = []
-      for item in objects:
-        if isinstance(item, bytes):
-          byte_objects.append(item)
-        elif isinstance(item, str):
-          byte_objects.append(item.encode(encoding))
-        else:
-          byte_objects.append(str(item).encode(encoding))
-      data = sep.join(byte_objects)
-      data += end
-      ttywrite(file, data)
-
-    try:
-      # Try to print with default encoding.
-      Py3Print(locale.getpreferredencoding(False))
-    except UnicodeEncodeError:
-      # If default failed, try to print with UTF-8.
-      Py3Print(UTF8)
-
-
-def ttywrite(fp, data):
+        expected_keywords[key] = value
+
+    return expected_keywords.values()
+
+  sep, end, file = _get_args(**kwargs)
+
+  sep = six.ensure_binary(sep)
+  end = six.ensure_binary(end)
+  byte_objects = [six.ensure_binary(item) for item in objects]
```

Comment:
You'll probably want to call `ensure_binary` on the string representation of the thing (whatever is returned from the object's `__str__` or `__repr__` methods).  Doing this means we can't really use a (simple) list comprehension, but would instead look more like this:

```
byte_objects = []
for item in objects:
  if not isinstance(item, (six.binary_type, six.text_type)):
    # If the item wasn't bytes or unicode, its __str__ method
    # should return one of those types.
    item = str(item)

  if isinstance(item, six.binary_type):
    byte_objects.append(item)
  else:
    # The item should be unicode. If it's not, ensure_binary()
    # will throw a TypeError.
    byte_objects.append(six.ensure_binary(item))
```

Reply:
That makes sense! I was previously doing `[six.ensure_binary(str(item)) ... ]` but removed the `str()` since it would fail in the case of binary/unicode mismatch. This looks closer to the original logic and makes sense why now, I'll edit this in.